sdk/storage/Azure.Storage.Files/src/FileUriBuilder.cs

```diff
@@ -119,6 +120,13 @@ public SasQueryParameters Sas
         }
         private SasQueryParameters _sas;
 
+        /// <summary>
+        /// Get the terminal piece from the <see cref="DirectoryOrFilePath"/>, or null if
+        /// not present in the <see cref="Uri"/>.
+        /// </summary>
+        internal string TerminalPiece =>
+            DirectoryOrFilePath.TrimEnd('/').Split('/').LastOrDefault();
```

Comment:
Nit - it'd be kind of nice to allocate a little less here by combining `LastIndexOf` and `Substring` but I don't think we need to take the time to do that now since this isn't a high traffic API.

Reply:
Yeah, since we cache the data in the client, we would only be taking the hit once for the lifetime of a client if the property is even used.

Reply:
Further thought - I don't trust your parsing code because it's too intelligible and I'm pretty sure any real path parsing is going to be an explosion of edge cases.  How about we use https://docs.microsoft.com/en-us/dotnet/api/system.uri.segments?view=netframework-4.8 instead?

Reply:
Okay, maybe what you have [would work fine,](https://github.com/dotnet/corefx/blob/master/src/System.Private.Uri/src/System/Uri.cs#L827) but I'd still like somebody else to get the bug reports if something's not quite right.  😄

Reply:
It looks like Segments would actually treat a slash as a segment in the case of double slashes:
```
var myBuilder = new FileUriBuilder(new Uri("http://www.sdafsdf.com/foo/bar//"));

Assert.AreEqual("bar", myBuilder.Uri.Segments.Last()); // fails (returns / as final segment)
Assert.AreEqual("bar", myBuilder.TerminalDirectoryOrFileName); // passes
```
Also, it is acting on the absolute path as opposed to the already parsed DirectoryOrFilePath.

Reply:
That's fair - I'm probably being overly paranoid because URIs aren't nearly as bad as file paths.  

Alright.